### ISSUE TITLE
No way to change when the plugin is run

### ISSUE THREAD

**Issue by daniel-brenot** [namics/webpack-concat-text-plugin#22]
Currently the plugin always runs in the emit stage. It would be desired to have the ability to specify that i want the plugin to run after the compilation is complete. This could be done through an option that defaults to emit.

**Comment by zidizei** [comment]
Thanks for opening the PR. Is there a reason why you need to customise this behaviour? Are you encountering a case where the current process is not working for you?

**Comment by daniel-brenot** [comment]
Yes. Currently i am using this to bundle multiple files together after another webpack process that produces assets, and i need to use this plugin to modify those assets. Currently this only triggers early in the webpack cycle, so the assets do not exist when i go to activate this plugin.

**Comment by zidizei** [comment]
To be honest, I don't feel comfortable exposing something that close the the webpack API via a configuration. However, there is no particular reason why this plugin is currently run during `emit`, so I'm considering to change it to something like `afterEmit`.

What exactly are those assets that are being generated, though? Shouldn't the responsible plugin be able to concat its output somehow?

**Comment by daniel-brenot** [comment]
Since the plugin that i'm using is what is the central runner of the build process, it only makes sense for it to run during the emit hook. The responsible plugin should be able to concat its output, but in my use case it has been unable to. For the moment, this plugin is the only way for me to do that. I'm not sure what the concern is of exposing something close to the webpack API like you mentioned though, it isn't exactly a security risk if the user is the one configuring the use of this library.

**Comment by daniel-brenot** [comment]
Changing it to afterEmit would work for my use case, but i cant guarantee that it will work for everyone. If a sizeable number of people use this plugin, changing it to afterEmit is bound to break some peoples builds.

**Comment by daniel-brenot** [comment]
Bumping to see if theres any update to this?

**Comment by zidizei** [comment]
Hi @daniel-brenot, apologies for the delays.

I still think it would be best to solve your issue outside of this plugin. If you need to concat assets that have been emitted by other plugins, have you tried re-arranging the order of the plugins in your webpack config? 

Also, if that plugin you mentioned should be able to concat its output on its own, but somehow cannot, it might be helpful to submit an issue for it as well.

**Comment by daniel-brenot** [comment]
I'm working with the angular cli, and with my use case this will never be fixed by them. Also, re-arranging the plugins won't change how they run, the plugin is the only way to decide when code is run in the stages. Your plugin will always be run before assets are emitted currently, and if that isn't fixed there is no way for me to continue using this moving forward. I have no way to fix this issue outside of the plugin because it is an issue with the plugin itself.

**Comment by daniel-brenot** [comment]
I ask that you seriously reconsider allowing controlling the point at which the plugin runs. It is a small change that already has a pr and doesn't have any risk associated with it. I'm sure i won't be the only one that benefits from this change. If not, then i will have to fork this for use in my workspace and i would really like to avoid that, especially since the change is so small.

**Comment by zidizei** [comment]
I believe plugin order does matter, though, so this plugin can pick up other emitted assets when run last.

In any case, I'd like to suggest the following changes to your PR:

- Instead of exposing all webpack hooks, add a toggle to the config to switch between `emit` and `afterEmit`. Feel free to name this setting the way you like it
- Extend the tests to take this new setting into account

**Comment by daniel-brenot** [comment]
It only matters in the sense that it affects when the plugin runs the hook specified. For example, if you register plugin a before plugin b, and they both have the 'afterEmit' hook tapped into, then plugin a will have it's afterEmit called first. However, if plugin b hooks into 'beforeEmit' for example, then it will be called first. In this case, you don't have any control over which one runs first.

I'm curious how having a flag is any different than using the bare hook name. The only difference functionally-wise between my solution and yours is that my solution allows greater control over when the plugin is run, and the one you propose only allows for having it execute before or during. If it's a worry of a user inputting an invalid value, I could always add input validation to ensure that the value is one of the valid webpack hooks.

For my use case I need afterEmit for the moment, however there is no guarantee that afterEmit or emit will cover the cases i will encounter in the future, and I don't want to run into that problem and have to come back and make another issue in a few months and request changing a flag to a string, requiring code that supports an older flag as well as a string, causing further confusion.

If you are worried about certain webpack hooks being used, we could also restrict the values to an array of hooks that could be used, that way if someone requested for more hooks to be added in the future, the list could easily be extended rather than the flag mechanism having to be changed to support an additional time to hook into. 

I'm not quite sure how I would test this feature. It's just a string argument passed into webpack, so as long as it is in the accepted range of values in the documentation it shouldn't have any behaviour that is out of the ordinary. If you know of a way to test this, feel free to mention it and I will add a test to check for it.

**Comment by daniel-brenot** [comment]
As an example of another plugin that does this, the file manager plugin lets you execute steps at specific parts, and so does the webpack shell plugin: 
https://github.com/s00d/webpack-shell-plugin-next

**Comment by zidizei** [comment]
> I'm curious how having a flag is any different than using the bare hook name. The only difference functionally-wise between my solution and yours is that my solution allows greater control over when the plugin is run, and the one you propose only allows for having it execute before or during. If it's a worry of a user inputting an invalid value, I could always add input validation to ensure that the value is one of the valid webpack hooks.

I don't want to expose webpack hooks to the end user, so I would prefer to map this with specific config values, like webpack-shell-plugin-next does with `onBeforeBuild`, etc. Not everyone might (want to) know which hooks are running when, so using specific configuration values should make it easier to understand.

> For my use case I need afterEmit for the moment, however there is no guarantee that afterEmit or emit will cover the cases i will encounter in the future, and I don't want to run into that problem and have to come back and make another issue in a few months and request changing a flag to a string, requiring code that supports an older flag as well as a string, causing further confusion.

I'm sorry but I'm would only allow an option to have this plugin either run before asset emission or after. If your use case is that complex, it might be easier to just run a separate script after your webpack build.

> I'm not quite sure how I would test this feature. It's just a string argument passed into webpack, so as long as it is in the accepted range of values in the documentation it shouldn't have any behaviour that is out of the ordinary. If you know of a way to test this, feel free to mention it and I will add a test to check for it.

I tried out your PR in a minimal setup, and it doesn't seem to work with `afterEmit`. Can you verify this?
In order to get this to run on that hook would require a different implementation (which should also be tested on its own), because currently, concatenated files are emitted as assets themselves.
